### PR TITLE
Monitoring mode + refactor

### PR DESCRIPTION
**Pull request by tom-diacono**

### CODE REVIEW

File: src/asteroid_sdk/api/llm_data_handler.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+)
+from src.asteroid_sdk.api.supervision_runner import SupervisionRunner
+
+
+class AsteroidLoggingError(Exception):
+    """Raised when there's an error logging to Asteroid API."""
+    pass
+
+
+class LLMDataHandler:
```
Comment (mlcocdav) [inline src/asteroid_sdk/api/llm_data_handler.py:20]: Can we think of better name for `LLMDataHandler`.  `APILogger` and `SupervisionRunner` are kind of clear what their functionality is from the name, but LLMDataHandler is not clear. My suggestion: `AsteroidChatSupervisionManager`. (Here chat relates to the openai chat endpoint we wrap around (which also accepts images))

File: run.py [hunk 2]
```diff
@@ -126,39 +126,46 @@ def book_hotel(location: str, checkin: str, checkout: str, maximum_price: float)
...
     messages.append({"role": "user", "content": user_input})
-    
+
     # Make API call
     response = wrapped_client.chat.completions.create(
         model="gpt-4o-mini",
         messages=messages,
         tools=tools,
         tool_choice="auto",
         temperature=1.5,
```
Comment (mlcocdav) [inline run.py:139]:
```suggestion
        temperature=0,
```

**Review by mlcocdav — COMMENTED** [review]
Looks great! I like the separation of concerns. I've tested it and it works as expected. I only have few minor comments.

Other point:
- We currently don't log results of the function call. We only see them if the user calls the model again and the result is in the context window. Example of finished execution below: 
image
The user finished execution but our UI does not show the function result. But since we  use the supervise() decorator around the tool call we actually have information about the execution. Would it be possible to add logging to the supervise() decorator in `src/asteroid_sdk/supervision/decorators.py#L45`? It might be needed to add endpoint to backend to just append one message to the chat @joehewett

File: src/asteroid_sdk/api/llm_data_handler.py (hunk 1 above)
Comment (tom-diacono) [reply src/asteroid_sdk/api/llm_data_handler.py:20]: Yes, 100%. I was thinking about this last night and funnily enough thought `Manager` fits in there perfectly... Then forgot this morning ;) Will add now

**Comment by tom-diacono** [comment]
@mlcocdav  Defo not quite there with separation IMO- I guess the goal was to make a start to it before it got more difficult- but there's defo more to be done :)

File: src/asteroid_sdk/api/supervision_runner.py [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+            f"Resample {resample+1}: {message['role']}: {message['content']}\n"
+            for resample, message in enumerate(updated_messages[1:])
+        ])
+        rejection_message = (
+            f"The agent requested to execute a function but it was rejected by some supervisors.\n"
+            f"We tried {n_resamples} times to get a valid response but it was rejected each time.\n"
+            f"Here is the feedback from the supervisors: \n{explanations}\n"
+            f"This is not a message from the user but from a supervisor system that is helping the agent to improve its behavior. You should try something else!"
+        )
+        # TODO - should we add a `Choice` in here- or just return the message? Currently, we're returning the message and
```
Comment (joehewett) [inline src/asteroid_sdk/api/supervision_runner.py:469]: I think ultimately we'll fit into the existing workflow better if we return a reject reason. Not vital right now

**Pull request merged**